virttest/qemu_devices/qdevices.py

```diff
@@ -359,6 +359,13 @@ def is_pcie_device(self):
                         "vhost-user-fs-pci"]
         return (driver in pcie_drivers or driver.startswith("virtio-"))
 
+    def _parameters_replace(self):
+        mapping = {"serial": ("NO_EQUAL_STRING", "on")}
+        for key, value in mapping.items():
+            origin_val = self.get_param(key, "")
+            if origin_val == value[0]:
+                self.set_param(key, value[1])
```

Comment:
I didn't get the point that we set the option `serial` to `on`. 
I check the comments about `NO_EQUAL_STRING`:
https://github.com/avocado-framework/avocado-vt/blob/247daa8b0bd3b9e718928e46dc35e4ad44fe09ff/virttest/qemu_vm.py#L514
```
                # "NO_EQUAL_STRING": set option as a option string only,
                #                    even without "=".
```
From my understanding, the `NO_EQUAL_STRING` is to handle the options without any values, not just for the `serial` option.
But I did not know how to handle such options in the JSON command line.

Reply:
> I didn't get the point that we set the option `serial` to `on`. I check the comments about `NO_EQUAL_STRING`:
> 
> https://github.com/avocado-framework/avocado-vt/blob/247daa8b0bd3b9e718928e46dc35e4ad44fe09ff/virttest/qemu_vm.py#L514
> 
> ```
>                 # "NO_EQUAL_STRING": set option as a option string only,
>                 #                    even without "=".
> ```
> 
> From my understanding, the `NO_EQUAL_STRING` is to handle the options without any values, 

yes, property only, no value in raw format

not just for the `serial` option. But I did not know how to handle such options in the JSON command line.

it looks only usb device's serial option uses the NO_EQUAL_STRING value now, @nickzhq , is that true?
for the value 'on' in json format, I found something in format_disk.py in tp-qemu:

If a device option(bool/str) in qemu cmd line doesn't have a value, qemu assigns the value as "on".

So, it looks that qemu always set 'on' for a property without any value.

Reply:
> > I didn't get the point that we set the option `serial` to `on`. I check the comments about `NO_EQUAL_STRING`:
> > https://github.com/avocado-framework/avocado-vt/blob/247daa8b0bd3b9e718928e46dc35e4ad44fe09ff/virttest/qemu_vm.py#L514
> > 
> > ```
> >                 # "NO_EQUAL_STRING": set option as a option string only,
> >                 #                    even without "=".
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > From my understanding, the `NO_EQUAL_STRING` is to handle the options without any values,
> 
> yes, property only, no value in raw format
> 
> not just for the `serial` option. But I did not know how to handle such options in the JSON command line.
> 
> it looks only usb device's serial option uses the NO_EQUAL_STRING value now, @nickzhq , is that true? for the value 'on' in json format, I found something in format_disk.py in tp-qemu:
> 
> If a device option(bool/str) in qemu cmd line doesn't have a value, qemu assigns the value as "on".
> 
> So, it looks that qemu always set 'on' for a property without any value.

Thanks to @zhencliu for the response.